Read and use deprecated configuration (as well as emitting a warning)

bossmc opened the pull request:
Original change written by @flip1995 I've simply rebased to master and fixed up the formatting/tests.  This change teaches the configuration parser which config key replaced a deprecated key and attempts to populate the latter from the former.  If both keys are provided this fails with a duplicate key error (rather than attempting to guess which the user intended).

Currently this on affects `cyclomatic-complexity-threshold` -> `cognitive-complexity-threshold` but will also be used in #8974 to handle `blacklisted-names` -> `disallowed-names`.

```
changelog: deprecated configuration keys are still applied as if they were provided as their non-deprecated name.
```

- [x] `cargo test` passes locally
- [x] Run `cargo dev fmt`

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Jarcho (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md) for more information.

flip1995 commented:
I only vaguely remember writing this :sweat_smile: Thanks for seeing it to an end!

Jarcho commented:
One small issue. This won't warn about duplicate fields when the old name appears after the new name. It's not a big deal as you still get the warning about the deprecated field.

Otherwise LGTM.

bossmc commented:
I've added a check to handle duplicates in either order and added a test for duplicates.  I also extended the error message to be clear what's happening if one of the duplicates is using the deprecated name (though this only works on one direction).

Jarcho commented:
Thank you. That should be good enough. The deprecation warning mentions both names in either case.

@bors r+

bors commented:
:pushpin: Commit ea25ef10cf942172e09b463c305b934fabccc8e2 has been approved by `Jarcho`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Jarcho ea25ef10cf942172e09b463c305b934fabccc8e2 -->
<!-- homu: {"type":"Approved","sha":"ea25ef10cf942172e09b463c305b934fabccc8e2","approver":"Jarcho","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit ea25ef10cf942172e09b463c305b934fabccc8e2 with merge 53a09d48557119d2b2a875d77a284b470a95e1c8...
<!-- homu: {"type":"BuildStarted","head_sha":"ea25ef10cf942172e09b463c305b934fabccc8e2","merge_sha":"53a09d48557119d2b2a875d77a284b470a95e1c8"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/7571200772?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/7571190626?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/7571317478?check_suite_focus=true)
Approved by: Jarcho
Pushing 53a09d48557119d2b2a875d77a284b470a95e1c8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Jarcho","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/7571317478?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/7571200772?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/7571190626?check_suite_focus=true"},"merge_sha":"53a09d48557119d2b2a875d77a284b470a95e1c8"} -->
